crates/rome_js_analyze/src/analyzers/nursery/no_redundant_use_strict.rs

```diff
@@ -0,0 +1,132 @@
+use crate::JsRuleAction;
+use rome_analyze::{context::RuleContext, declare_rule, ActionCategory, Ast, Rule, RuleDiagnostic};
+use rome_console::markup;
+use rome_diagnostics::Applicability;
+use rome_js_syntax::{AnyJsRoot, JsDirective, JsDirectiveList, JsFunctionBody};
+
+use rome_rowan::{AstNode, BatchMutationExt};
+
+declare_rule! {
+ /// Prevents from having redundant \"use strict\"
+ ///
+ /// ## Examples
+ ///
+ /// ### Invalid
+ /// "use strict";
+ /// function foo() {
+ ///  	"use strict";
+ /// }
+ ///
+ /// ### valid
+ /// "use strict";
+ ///
+ /// function foo() {
+ ///
+ /// }
+ ///
+ /// ### valid
+ ///
+ ///  function foo() {
+ ///     "use strict";
+ /// }
+ /// function bar() {
+ ///     "use strict";
+ /// }
+ ///
+ ///
+
+ pub(crate) NoRedundantUseStrict {
+     version: "11.0.0",
+     name: "NoRedundantUseStrict",
+     recommended: false,
+    }
+}
+
+pub struct State {
+    first: JsDirective,
+    redundant: Vec<JsDirective>,
+}
+
+impl Rule for NoRedundantUseStrict {
+    type Query = Ast<JsDirective>;
+    type State = State;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let mut directives = vec![(*node).clone()];
+
+        let check_use_strict_directive =
+            |node: &JsDirective| node.value_token().map(|v| v.text_trimmed().to_owned());
+
+        // We have to trace the check from great-grandfather's ancestors to skip itself.
+        let ancestors = node.syntax().parent()?.parent()?.parent()?.ancestors();
```

Comment:
You can use `ancestors()` directlly and then match on the kind and I then recommend you to declare a node `AnyNodeWithDirevices`

```
declare_node_union! {
  AnyNodeWithDirevices = JsFunctionBody | AnyJsRoot
}

impl AnyNodeWithDirevices {
 fn directives(&self) -> JsDirectiveList {
	match self {
		...
    }
  }
}
```

I also recommend generating a different diagnostic for every redundant directive rather than aggregating them together. 

You can then write

```suggestion
let mut outer_most: Option<JsDirective> = None;

for parent in node.syntax().ancestors().skip(1).filter_map(AnyNodeWithDirevices::cast) {
	for directive in parent.directives() {
		if directive.value_token().map_or(false, |t| matches!(t.text_trimmed(), "'use strict'" | "\"use strict\"")) {
			outer_most = Some(directive);
			break; // continue with next parent
		}
	 }
} 
```

Reply:
Additionally we may not necessarily want to skip the `JsDirectiveList` that contains the query node, we could also check it for duplicates to catch a case like this:
```js
"use strict";
"use strict";
```
I'm not sure that's really necessary though, it seems like an unlikely edge case ...

Reply:
Thanks @MichaReiser, I was not aware of `declare_node_union!`. I will update the code based on your suggestion!